transport/awssqs/consumer.go

```diff
@@ -0,0 +1,242 @@
+
+	if len(c.finalizer) > 0 {
+		defer func() {
+			for _, f := range c.finalizer {
+				f(ctx, &msgs)
+			}
+		}()
+	}
+
+	for _, f := range c.before {
+		ctx = f(ctx, &msgs)
+	}
+
+	for _, msg := range msgs {
+		if err := c.HandleSingleMessage(ctx, msg, &leftMsgs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// HandleSingleMessage handles a single sqs message
+func (c Consumer) HandleSingleMessage(ctx context.Context, msg *sqs.Message, leftMsgs *[]*sqs.Message) error {
+	req, err := c.dec(ctx, msg)
+	if err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	response, err := c.e(ctx, req)
+	if err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	responseMsg := sqs.SendMessageInput{}
+	for _, f := range c.after {
+		ctx = f(ctx, msg, &responseMsg, leftMsgs)
+	}
+
+	if !c.wantRep(ctx, msg) {
+		// Message does not expect answer
+		return nil
+	}
+
+	if err := c.enc(ctx, &responseMsg, response); err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	if _, err := c.sqsClient.SendMessageWithContext(ctx, &responseMsg); err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+	return nil
+}
+
+// ErrorEncoder is responsible for encoding an error to the consumer reply.
+// Users are encouraged to use custom ErrorEncoders to encode errors to
+// their replies, and will likely want to pass and check for their own error
+// types.
+type ErrorEncoder func(ctx context.Context, err error, req *sqs.Message, sqsClient SQSClient)
+
+// ConsumerFinalizerFunc can be used to perform work at the end of a request
+// from a publisher, after the response has been written to the publisher. The
+// principal intended use is for request logging.
+// Can also be used to delete messages once fully proccessed
+type ConsumerFinalizerFunc func(ctx context.Context, msg *[]*sqs.Message)
+
+// DefaultErrorEncoder simply ignores the message. It does not reply
+// nor Ack/Nack the message.
+func DefaultErrorEncoder(context.Context, error, *sqs.Message, SQSClient) {
+}
+
+// DoNotExtendVisibilityTimeout is the default value for visibilityTimeoutFunc
+// It returns no error and does nothing
+func DoNotExtendVisibilityTimeout(context.Context, SQSClient, *string, int64, *[]*sqs.Message) error {
+	return nil
+}
+
+// EncodeJSONResponse marshals response as json and loads it into input MessageBody
+func EncodeJSONResponse(_ context.Context, input *sqs.SendMessageInput, response interface{}) error {
+	payload, err := json.Marshal(response)
+	if err != nil {
+		return err
+	}
+	input.MessageBody = aws.String(string(payload))
+	return nil
+}
+
+// SQSClient is an interface to make testing possible.
+// It is highly recommended to use *sqs.SQS as the interface implementation.
+type SQSClient interface {
```

Comment:
This name stutters a bit: awssqs.SQSClient. Consider `Client`.

Reply:
I agree